odl/discr/partition.py

```diff
@@ -0,0 +1,561 @@
+
+"""Partitons of interval products based on tensor grids.
+
+A partition of a set is a finite collection of nonempty, pairwise
+disjoint subsets whose union is the original set. The partitions
+considered here are based on hypercubes, i.e. the tensor products
+of partitions of intervals.
+"""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+
+from future import standard_library
+standard_library.install_aliases()
+from builtins import object, range, super, zip
+
+# External module imports
+import numpy as np
+
+# ODL imports
+from odl.discr.grid import TensorGrid, RegularGrid
+from odl.set.domain import IntervalProd
+from odl.util.utility import array1d_repr
+
+
+__all__ = ('RectPartition', 'uniform_partition')
+
+_POINT_POSITIONS = ('center', 'left')
+
+
+class RectPartition(object):
+
+    """Rectangular partition by hypercubes based on `TensorGrid`s.
+
+    In 1d, a partition of an interval is implicitly defined by a
+    collection of points x[0], ..., x[N-1] (a grid) which are chosen to
+    lie in the center of the subintervals. The i-th subinterval is thus
+    given by
+    ::
+        I[i] = [(x[i-1]+x[i])/2, (x[i]+x[i+1])/2]
+    """
+
+    def __init__(self, grid, begin=None, end=None, **kwargs):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        grid : `TensorGrid`
+            Spatial points supporting the partition
+        begin, end : array-like, shape ``(grid.ndim,)``
+            Spatial points defining the begin and end of an interval
+            product to be partitioned. ``begin`` must be component-wise
+            at most ``grid.min_pt``, and ``end`` at least
+            ``grid.max_pt``. If not provided, they are inferred from
+            the grid.
+        begin_axes, end_axes : sequence of `int`, optional
+            Dimensions in which to apply ``begin`` and ``end``. If
+            given, only these dimensions in ``begin`` or ``end``, resp.,
+            are considered - other entries are ignored.
+
+        Notes
+        -----
+        The endpoints ``begin`` and ``end`` are determined from the grid
+        as follows::
+
+            begin = x[0] - (x[1] - x[0]) / 2
+            end = x[N-1] + (x[N-1] - x[N-2]) / 2
+        """
+        if not isinstance(grid, TensorGrid):
+            raise TypeError('{!r} is not a TensorGrid instance.'
+                            ''.format(grid))
+
+        # Begin and end: check correct shapes and values if given,
+        # calculate otherwise.
+        begin_axes = kwargs.pop('begin_axes', None)
+        end_axes = kwargs.pop('end_axes', None)
+
+        begin = self._calc_begin_end(grid, begin_axes, begin, 'begin')
+        end = self._calc_begin_end(grid, end_axes, end, 'end')
+
+        super().__init__()
+        self._bbox = IntervalProd(begin, end)
+        self._grid = grid
+
+    def _calc_begin_end(self, grid, vec_axes, vector, which):
```

Comment:
This functions logic is quite complicated for something that should be quite simple (at least from its name), it could use some simplifications.